src/client/dfs/dfs.c

```diff
@@ -699,11 +699,39 @@ open_file(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
 		return ENOENT;
 
 fopen:
-	if (!S_ISREG(entry.mode)) {
-		if (entry.value) {
-			D_ASSERT(S_ISLNK(entry.mode));
+	if (S_ISLNK(entry.mode)) {
+		/* conditionally dereference symbolic links */
+		if (entry.value == NULL) {
+			D_ERROR("Null Symlink value\n");
+			return EIO;
+		}
+
+		if (flags & O_NOFOLLOW) {
 			D_FREE(entry.value);
+			return ELOOP;
 		}
+
+		dfs_obj_t *sym;
+
+		rc = dfs_lookup(dfs, entry.value, flags, &sym, NULL, NULL);
+		if (rc) {
+			D_FREE(entry.value);
```

Comment:
at this point you can free entry.value without checking for error condition and avoid doing it again below.

Reply:
Yes, sorry. mpifileutils has different style guidelines :)